webapp/loggedin/modules/analysis/processingChain/processingChainState.js

```diff
@@ -15,34 +16,47 @@ export const getState = R.pipe(AnalysisState.getState, R.prop(stateKey))
 
 // ====== READ
 
-const getProcessingChainOrig = R.pipe(getState, R.propOr({}, keys.orig))
+const _getStateProp = (key, defaultValue = null) => R.pipe(getState, R.propOr(defaultValue, key))
 
-export const getProcessingChain = R.pipe(getState, R.propOr({}, keys.dirty))
+const getProcessingChainOrig = _getStateProp(keys.orig, {})
+
+export const getProcessingChain = _getStateProp(keys.dirty, {})
 
 // ====== UPDATE
 
 export const assocProcessingChain = chain => R.pipe(R.assoc(keys.orig, chain), R.assoc(keys.dirty, chain))
 
-const _assocProcessingChainProp = (key, fn) => state =>
+const _updateChain = (key, fn) => state =>
   R.pipe(R.prop(key), fn, chainUpdated => R.assoc(key, chainUpdated)(state))(state)
 
-const _assocProcessingChainPropOrig = fn => _assocProcessingChainProp(keys.orig, fn)
+const _updateChainOrig = fn => _updateChain(keys.orig, fn)
 
-const _assocProcessingChainPropDirty = fn => _assocProcessingChainProp(keys.dirty, fn)
+const _updateChainDirty = fn => _updateChain(keys.dirty, fn)
 
-export const assocPropDirty = (key, value) => _assocProcessingChainPropDirty(ProcessingChain.assocProp(key, value))
+export const assocPropDirty = (key, value) => _updateChainDirty(ProcessingChain.assocProp(key, value))
 
 export const assocProcessingSteps = steps =>
   R.pipe(
-    _assocProcessingChainPropDirty(ProcessingChain.assocProcessingSteps(steps)),
-    _assocProcessingChainPropOrig(ProcessingChain.assocProcessingSteps(steps)),
+    _updateChainDirty(ProcessingChain.assocProcessingSteps(steps)),
+    _updateChainOrig(ProcessingChain.assocProcessingSteps(steps)),
   )
 
-export const mergeDirty = state => {
-  const chainDirty = R.prop(keys.dirty, state)
-  return R.assoc(keys.orig, chainDirty, state)
+export const appendProcessingStep = step => _updateChainDirty(ProcessingChain.assocProcessingStep(step))
+
+export const dissocStepTemporary = state => {
+  const steps = R.pipe(R.prop(keys.dirty), ProcessingChain.getProcessingSteps)(state)
+  const stepLast = R.last(steps)
+
+  return ProcessingStep.isTemporary(stepLast)
+    ? R.assocPath([keys.dirty, ProcessingChain.keys.processingSteps], R.dropLast(1, steps))(state)
```

Comment:
maybe it's better to not export ProcessingChain.keys and use ProcessingChain.assocProcessingSteps here

Reply:
i agree, let's do it in the next pull request?

Reply:
ok